transaction_verifier.go

```diff
@@ -0,0 +1,148 @@
+// Since each bit of the index is mapped to an element in the resulting positions array,
+// the index must contain a number of bits amounting to the depth parameter, which means it must be smaller than 2 ^ depth.
+func (t *TransactionVerifier) getVectorCommitmentPositions(index uint64, depth uint64) ([]Position, error) {
+	if depth == 0 {
+		return []Position{}, ErrInvalidTreeDepth
+	}
+
+	if index >= 1<<depth {
+		return []Position{}, ErrIndexDepthMismatch
+	}
+	directions := make([]Position, depth)
+	for i := len(directions) - 1; i >= 0; i-- {
+		directions[i] = Position(index & 1)
+		index >>= 1
+	}
+
+	return directions, nil
+}
+
+func (t *TransactionVerifier) computeVectorCommitmentMerkleRoot(leaf stateprooftypes.GenericDigest, leafIndex uint64, proof []byte, treeDepth uint64, hashFunc hash.Hash) (stateprooftypes.GenericDigest, error) {
+	if len(proof) == 0 && treeDepth == 0 {
+		return leaf, nil
+	}
+
+	nodeSize := uint64(hashFunc.Size())
+	if treeDepth*nodeSize != uint64(len(proof)) {
+		return stateprooftypes.GenericDigest{}, ErrProofLengthTreeDepthMismatch
+	}
+
+	positions, err := t.getVectorCommitmentPositions(leafIndex, treeDepth)
+	if err != nil {
+		return stateprooftypes.GenericDigest{}, err
+	}
+
+	currentNodeHash := leaf
+	for currentLevel := uint64(0); currentLevel < treeDepth; currentLevel++ {
+		siblingIndex := currentLevel * nodeSize
+		siblingHash := proof[siblingIndex : siblingIndex+nodeSize]
+
+		nextNode := []byte(stateprooftypes.MerkleArrayNode)
+		switch positions[currentLevel] {
+		case left:
+			nextNode = append(append(nextNode, currentNodeHash...), siblingHash...)
+		case right:
+			nextNode = append(append(nextNode, siblingHash...), currentNodeHash...)
+		default:
+			return []byte{}, ErrInvalidPosition
+		}
+
+		currentNodeHash = stateprooftypes.HashBytes(hashFunc, nextNode)
+	}
+
+	return currentNodeHash, nil
+}
+
+// VerifyTransaction receives a hashed transaction, a proof to compute the transaction's commitment, a proof
+// to compute the light block header's commitment, and an expected commitment to compare to.
+// Verification works as follows:
+//	1. Compute the transaction's vector commitment using the provided transaction proof.
+//	2. Build a candidate light block header using the computed vector commitment.
+// 	3. Compute the candidate light block header's vector commitment using the provided light block header proof.
+// 	4. Verify that the computed candidate vector commitment matches the expected vector commitment.
+func (t *TransactionVerifier) VerifyTransaction(transactionId types.Digest, transactionProofResponse models.ProofResponse, lightBlockHeaderProofResponse models.LightBlockHeaderProof, confirmedRound types.Round, seed stateprooftypes.Seed, blockIntervalCommitment types.Digest) error {
+	hashFunc, err := stateprooftypes.UnmarshalHashFunc(transactionProofResponse.Hashtype)
+	if err != nil {
+		return err
+	}
+
+	var stibHashDigest types.Digest
```

Comment:
if we are using genericDigest then we can just reference that to be a generic digest an not copy